### PR TITLE
Added C API bindings and bug fix related to CMFD tallies

### PR DESCRIPTION
**Pull request by shikhar413**
## Overview
This simple PR is part of a larger effort to move the existing CMFD code in the repository from Fortran to Python through the C API, and addresses some of the additional C API bindings necessary for this conversion. In addition, this PR fixes a bug related to existing Fortran CMFD code, where CMFD tallies cannot be accessed through the C API. These additions will allow cmfd_input.F90 be rewritten entirely through Python, the framework for which can be found on my branch `shikhar413::cmfd-capi`

## C API Bindings

The additional bindings introduced are getter and setter functions for the "estimator" and "type" of the Fortran tally object. The reason I include these functions are because tallies allocated through the `openmc_tally_set_type` function are tracklength estimators of type "volume". However, these attributes cannot be modified through Python alone without these bindings. In the case of CMFD tallies, a combination of mesh surface and volume types are used and they are all collision estimators. Using the default values when allocating CMFD tallies through Python will not replicate the desired behavior of these tallies.

### Issue

Ideally, I would have liked to combine `openmc_tally_set_type` and `openmc_tally_update_type` into a single method that passes in the type of interest as an optional argument (for example `new_tally = openmc.capi.Tally(type=...)`). However, additional arguments cannot be done in a straightforward way without changing the behavior of other classes, since the Tally constructor in `openmc/capi/tally.py` is an inherited class. As a result, users will have to create the tally with `openmc_tally_set_type`,  and change the type separately with `openmc_tally_update_type`. I'm thinking I should rename the function that's currently defined as `openmc_tally_set_type` to `openmc_tally_allocate` and `openmc_tally_update_type` to `openmc_tally_set_type`, so that the setter and getter functions are defined consistently like the others. 

## Bug Fix

Currently, CMFD tallies are allocated separately from tallies defined in tallies.xml. Instead, they are allocated in the `create_cmfd_tally` subroutine in `cmfd_input.F90`. While this does not affect the behavior of the CMFD algorithm as a whole, when using the C API bindings to inspect the behavior of CMFD tallies in between batches, the dictionary of tallies does not display properly. As an example, running the following commands in Python in `tests/regression_tests/cmfd_feed` results in an error:

```
>>> import openmc.capi as capi
>>> capi.init()
>>> capi.simulation_init()
>>> capi.next_batch()
>>> capi.tallies
```

The error relates to specific tallies not being found, and has to do with the fact that tallies created outside of input xml files are never added to `tally_dict` in Fortran. This PR fixes this issue by requiring `openmc_tally_set_id ` be called on CMFD tallies, which will add the tally to tally_dict. Since CMFD tallies don't have a user-defined id, an id is assigned to these tallies with `openmc_get_tally_next_id`, similar to how id's are assigned to CMFD filters.

## Miscellaneous

* The "reset" attribute of Fortran tally objects is removed, since it is not currently used anywhere in the code.
* A `reset` method is included in `openmc/capi/tally.py` that calls `_dll.openmc_tally_reset` to reset individual tallies through Python

### CODE REVIEW

File: openmc/capi/tally.py [hunk 1]
```diff
@@ -302,6 +351,9 @@ def std_dev(self):
 
         return std_dev
 
+    def reset(self):
```
Comment (paulromano) [inline openmc/capi/tally.py]: Add a docstring after this

File: openmc/capi/tally.py [hunk 2]
```diff
@@ -190,6 +215,30 @@ def active(self):
...
+    def type(self, type):
+        _dll.openmc_tally_update_type(self._index, type.encode())
+
+    @property
+    def estimator(self):
+        estimator = c_int32()
+        try:
+            _dll.openmc_tally_get_estimator(self._index, estimator)
+        except AllocationError:
+            return ""
```
Comment (paulromano) [inline openmc/capi/tally.py]: I would actually just let the exception be raised in this case (i.e., no need for try/except).

File: src/input_xml.F90 [hunk 3]
```diff
@@ -2169,15 +2170,15 @@ subroutine read_tallies_xml()
 
       ! Copy tally id
       if (check_for_node(node_tal, "id")) then
-        call get_node_value(node_tal, "id", t % id)
+        call get_node_value(node_tal, "id", tally_id)
```
Comment (paulromano) [inline src/input_xml.F90]: If you move the call to `openmc_tally_set_id` immediately after this, then `t % id` can be used as is below.

File: src/tallies/tally_header.F90 [hunk 4]
```diff
@@ -942,4 +1013,43 @@ function openmc_tally_set_scores(index, n, scores) result(err) bind(C)
...
+    if (index >= 1 .and. index <= size(tallies)) then
+      select case (type_)
+      case ('volume')
+        tallies(index) % obj % type = TALLY_VOLUME
+      case ('mesh-surface')
+        tallies(index) % obj % type = TALLY_MESH_SURFACE
+      case ('surface')
+        tallies(index) % obj % type = TALLY_SURFACE
+      case default
+        err = E_UNASSIGNED
```
Comment (paulromano) [inline src/tallies/tally_header.F90]: Use `E_INVALID_ARGUMENT` instead

File: src/tallies/tally_header.F90 [hunk 5]
```diff
@@ -656,6 +695,37 @@ function openmc_tally_results(index, ptr, shape_) result(err) bind(C)
...
+    if (index >= 1 .and. index <= size(tallies)) then
+      select case (estimator_)
+      case ('analog')
+        tallies(index) % obj % estimator = ESTIMATOR_ANALOG
+      case ('tracklength')
+        tallies(index) % obj % estimator = ESTIMATOR_TRACKLENGTH
+      case ('collision')
+        tallies(index) % obj % estimator = ESTIMATOR_COLLISION
+      case default
+        err = E_UNASSIGNED
```
Comment (paulromano) [inline src/tallies/tally_header.F90]: Use `E_INVALID_ARGUMENT`

**Comment by shikhar413** [comment]
Thanks for the review, I went ahead and changed the C API binding names accordingly. I had to move `openmc_tally_set_id ` right after the check to make sure duplicate id's weren't found in tally_dict so that the check wouldn't fail

**Comment by paulromano** [comment]
Come to think of it, the check for duplicate IDs should probably just be part of `openmc_tally_set_id` itself.

**Comment by shikhar413** [comment]
I tried doing that, but I am not sure if setting the error in tally_header and throwing it in input_xml is the correct way to do it?

**Comment by shikhar413** [comment]
The capi tests in Travis are failing, I need to look into this

**Pull request merged**